main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Completer.scala

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2023 Lukas Rønn
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.api.lsp.provider.completion
+
+import ca.uwaterloo.flix.api.lsp.{CompletionItem, Index}
+import ca.uwaterloo.flix.language.ast.TypedAst
+
+/**
+  * A common super-type for completers.
+  */
+sealed trait Completer {
+  /**
+    * Returns a List of LSP completion items for `this`.
+    */
+  def getCompletions(implicit context: CompletionContext, index: Index, root: TypedAst.Root): Iterable[CompletionItem] = this match {
```

Comment:
This could be an abstract method. Like an interface.

Then you just have `BuiltinTypeCompletionProvider ` and friends implement it?

Reply:
Makes sense?

Reply:
I'll work on it! 👍